File: frontend/src/types/shipment.ts
/**
 * SwiftDrop :: TypeScript Type Definitions
 * These mirror the Pydantic response schemas from the FastAPI backend.
 * Single source of truth for data shapes across the frontend (DRY principle).
 */

export type ShipmentStatus =
  | 'PENDING'
  | 'ASSIGNED'
  | 'PICKED_UP'
  | 'IN_TRANSIT'
  | 'DELIVERED';

export type ShipmentType = 'STANDARD' | 'EXPRESS' | 'FRAGILE';

export type UserRole = 'VENDOR' | 'DRIVER' | 'ADMIN';

// ── API Response Shapes ────────────────────────────────────────────────────────

export interface Shipment {
  id: string;
  vendor_id: string;
  driver_id: string | null;
  shipment_type: ShipmentType;
  status: ShipmentStatus;
  pickup_address: string;
  delivery_address: string;
  package_weight_kg: number;
  package_description: string | null;
  estimated_fare: number;
  final_fare: number | null;
  created_at: string;     // ISO-8601 datetime string from backend
  accepted_at: string | null;
  picked_up_at: string | null;
  delivered_at: string | null;
  updated_at: string;
}

export interface TokenResponse {
  access_token: string;
  token_type: string;
  role: UserRole;
  user_id: string;
}

// ── API Request Shapes ─────────────────────────────────────────────────────────

export interface BookShipmentRequest {
  shipment_type: ShipmentType;
  pickup_address: string;
  delivery_address: string;
  package_weight_kg: number;
  package_description?: string;
}

export interface LoginRequest {
  email: string;
  password: string;
}

// ── UI State Shapes ────────────────────────────────────────────────────────────

export interface AuthState {
  token: string | null;
  role: UserRole | null;
  userId: string | null;
  isAuthenticated: boolean;
}

File: backend/schemas/shipment.py
"""
SwiftDrop :: Shipment Schemas (Pydantic v2)
Role : Data Transfer Objects (DTOs) — I/O contracts for the API.
       Completely decoupled from the DB layer (Low Coupling).
"""
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Literal, Optional
from uuid import UUID

from pydantic import BaseModel, Field, field_validator


# ── Request Schemas ────────────────────────────────────────────────────────────

class ShipmentBookRequest(BaseModel):
    """Payload for POST /api/shipments/book (Vendor only)."""

    shipment_type: Literal["STANDARD", "EXPRESS", "FRAGILE"] = "STANDARD"
    pickup_address: str = Field(..., min_length=5, max_length=500)
    delivery_address: str = Field(..., min_length=5, max_length=500)
    package_weight_kg: Decimal = Field(..., gt=0, decimal_places=3)
    package_description: Optional[str] = Field(None, max_length=500)

    @field_validator("shipment_type", mode="before")
    @classmethod
    def normalize_type(cls, v: str) -> str:
        return v.upper().strip()


# ── Response Schemas ────────────────────────────────────────────────────────────

class ShipmentResponse(BaseModel):
    """
    API response for a shipment resource.
    model_config: from_attributes=True allows construction from
    asyncpg.Record objects via ShipmentResponse.model_validate(dict(row)).
    """

    id: UUID
    vendor_id: UUID
    driver_id: Optional[UUID] = None
    shipment_type: str
    status: str
    pickup_address: str
    delivery_address: str
    package_weight_kg: Decimal
    package_description: Optional[str] = None
    estimated_fare: Decimal
    final_fare: Optional[Decimal] = None
    created_at: datetime
    accepted_at: Optional[datetime] = None
    picked_up_at: Optional[datetime] = None
    delivered_at: Optional[datetime] = None
    updated_at: datetime

    model_config = {"from_attributes": True}


# ── Auth Schemas ────────────────────────────────────────────────────────────────

class LoginRequest(BaseModel):
    email: str = Field(..., min_length=5)
    password: str = Field(..., min_length=6)


class TokenResponse(BaseModel):
    access_token: str
    token_type: str = "bearer"
    role: str
    user_id: str
